daemon/containerd/image_history.go

```diff
@@ -2,14 +2,104 @@ package containerd
 
 import (
 	"context"
-	"errors"
+	"encoding/json"
 
+	"github.com/containerd/containerd/content"
+	containerdimages "github.com/containerd/containerd/images"
+	cplatforms "github.com/containerd/containerd/platforms"
+	"github.com/docker/distribution/reference"
 	imagetype "github.com/docker/docker/api/types/image"
-	"github.com/docker/docker/errdefs"
+	"github.com/docker/docker/pkg/platforms"
+	"github.com/opencontainers/image-spec/identity"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
 )
 
-// ImageHistory returns a slice of ImageHistory structures for the specified
-// image name by walking the image lineage.
+// ImageHistory returns a slice of HistoryResponseItem structures for the
+// specified image name by walking the image lineage.
 func (i *ImageService) ImageHistory(ctx context.Context, name string) ([]*imagetype.HistoryResponseItem, error) {
-	return nil, errdefs.NotImplemented(errors.New("not implemented"))
+	desc, err := i.resolveDescriptor(ctx, name)
+	if err != nil {
+		return nil, err
+	}
+
+	cs := i.client.ContentStore()
+	// TODO: pass the platform from the cli
+	conf, err := containerdimages.Config(ctx, cs, desc, platforms.AllPlatformsWithPreference(cplatforms.Default()))
+	if err != nil {
+		return nil, err
+	}
+
+	diffIDs, err := containerdimages.RootFS(ctx, cs, conf)
+	if err != nil {
+		return nil, err
+	}
+
+	blob, err := content.ReadBlob(ctx, cs, conf)
+	if err != nil {
+		return nil, err
+	}
+
+	var image ocispec.Image
+	if err := json.Unmarshal(blob, &image); err != nil {
+		return nil, err
+	}
+
+	history := []*imagetype.HistoryResponseItem{}
+
+	sizes := []int64{}
+	s := i.client.SnapshotService(i.snapshotter)
+
+	for i := range diffIDs {
+		diffIDs := diffIDs[0 : i+1]
+		chainID := identity.ChainID(diffIDs).String()
+
+		use, err := s.Usage(ctx, chainID)
+		if err != nil {
+			return nil, err
+		}
+
+		sizes = append(sizes, use.Size)
+	}
+
+	for _, h := range image.History {
+		size := int64(0)
+		if !h.EmptyLayer {
+			if len(sizes) == 0 {
+				return nil, errors.New("unable to find the size of the layer")
+			}
+			size = sizes[0]
+			sizes = sizes[1:]
+		}
+
+		history = append([]*imagetype.HistoryResponseItem{{
+			ID:        "<missing>",
+			Comment:   h.Comment,
+			CreatedBy: h.CreatedBy,
+			Created:   h.Created.Unix(),
+			Size:      size,
+			Tags:      []string{},
```

Comment:
```suggestion
```

Reply:
`Tags` is required in the API https://docs.docker.com/engine/api/v1.41/#tag/Image/operation/ImageHistory, the current implementation does indeed return `"Tags": null` but that just feels wrong for two reasons:
* there is no need for a slice to be nil, an empty slice avoids having to check  if it's null
* the field is required

Also, from [[here]()](https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices):
```
Note that there are limited circumstances where a non-nil but zero-length slice is preferred, such as when encoding JSON objects (a nil slice encodes to null, while []string{} encodes to the JSON array []).
```

Reply:
> the current implementation does indeed return `"Tags": null` but that just feels wrong for two reasons:
> 
> * there is no need for a slice to be nil, an empty slice avoids having to check  if it's null

While I agree with you in general, that returning an empty JSON array is more friendly to non-Go clients than `null`, it's rather futile to fix it here and now _because_ of the existing behaviour. Any client which wants to be compatible with any daemon older than v-next needs to be prepared to handle responses containing `"Tags": null`. And it's possible that some client could be using `history.Tags != null` to check if the image has any tags. I would love to make the API more consistent, but for the sake of forwards and backwards compatibility I think this change in behaviour needs to be API version-gated. (Furthermore, promoting the `nil` slice to an empty slice before marshaling to JSON depending on the API version is well within the scope of responsibility for an API handler function, IMO.)

In summary, my recommended course of action is to:
1. change this implementation to return `Tags: nil` to match the existing behaviour, and
2. follow-up with a PR which changes `(*imageRouter).getImagesHistory` to respond with `Tags: []` when API version >= 1.44

Reply:
done!

I opened a ticket https://github.com/moby/moby/issues/44872